### PR TITLE
Standalone Lite: Move apply_func utilities

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Moves apply_func utils to Lite standalone package. Requested by reviewer to split out of #14492 

Needs to be merged with rebase+merge.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
No. Splitting out changes from a PR that is already splitting out changes from another PR already splitting out changes from the original PR is not fun. 

cc @borda @tchaton @rohitgr7 @carmocca @justusschock @awaelchli @akihironitta

### CODE REVIEW

File: src/lightning_lite/utilities/apply_func.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    # https://github.com/pytorch/pytorch/blob/v1.8.1/torch/nn/parallel/scatter_gather.py#L4-L8
+    return isinstance(obj, tuple) and hasattr(obj, "_asdict") and hasattr(obj, "_fields")
+
+
+def _is_dataclass_instance(obj: object) -> bool:
+    # https://docs.python.org/3/library/dataclasses.html#module-level-decorators-classes-and-functions
+    return dataclasses.is_dataclass(obj) and not isinstance(obj, type)
+
+
+def apply_to_collection(
```
Comment (carmocca) [inline src/lightning_lite/utilities/apply_func.py:61]: Opened https://github.com/Lightning-AI/utilities/pull/32 for this

We just need to release it to PyPI

Comment (awaelchli) [reply src/lightning_lite/utilities/apply_func.py:61]: Can it be done in parallel? I don't think it needs to block this. There will be a CI PR needed to add this as dependency and so on, which can take some time.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/utilities/apply_func.py [hunk 2]
```diff
@@ -13,327 +13,70 @@
...
-    return torch.from_numpy(value).to(device)
-
-
-CONVERSION_DTYPES: List[Tuple[Any, Callable[[Any, Any], Tensor]]] = [
-    # bool -> uint8 as bool -> torch.bool triggers RuntimeError: Unsupported data type for NCCL process group
-    (bool, partial(to_dtype_tensor, dtype=torch.uint8)),
-    (int, partial(to_dtype_tensor, dtype=torch.int)),
-    (float, partial(to_dtype_tensor, dtype=torch.float)),
-    (np.ndarray, from_numpy),
-]
```
Comment (carmocca) [inline src/pytorch_lightning/utilities/apply_func.py:48]: Just noticed we didn't deprecate this. But it's very tricky to do, being a global list.

Should we try or yolo?

Comment (awaelchli) [reply src/pytorch_lightning/utilities/apply_func.py:48]: It's going to far imo :)

**Pull request merged**